### PR TITLE
Provide interface for progress bars in `@thread_workers`

### PR DESCRIPTION
**Pull request by DragaDoncila**
# Description
- Add `progress` dictionary as argument to `@thread_worker` to support updating progress bar on yields
- Add increment methods to `progress`
- Users can choose to turn progress bar into an indeterminate one if the number of expected `yields` is exceeded
- Added tests and minimal example

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
Must merge #2580 first

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by DragaDoncila** [comment]
@tlambert03 I think this is ready for a first pass when you get a chance. I've kept it pretty minimal for now.

File: napari/_qt/qthreading.py
Comment (goanpeca) [inline napari/_qt/qthreading.py]: `trans._("_progress argument was passed but worker is not GeneratorWorker", deferred=True)`

**Review by goanpeca — APPROVED** [review]
Thanks for working on this @DragaDoncila. All looks good :) 

One missing translation ;-)

**Review by sofroniewn — APPROVED** [review]
LGTM! Also looks like all tests are passing 🚀

I'll wait for @tlambert03 to review before merge though

File: examples/progress_bar_threading.py
Comment (tlambert03) [inline examples/progress_bar_threading.py]: my first inclination would have been to just do this by default (no option provided)... That is, if worker is still yielding even when it said it should be done, I'd imagine we want to show that somehow... and the indeterminate state seemed like a good way to do that.  Did you run into a scenario that made you think that should not be the default behavior?

File: napari/_qt/qthreading.py
Comment (tlambert03) [inline napari/_qt/qthreading.py]: providing an empty dict (which is Falsy) feels a little funny as a way indicate "yes, please give me a progress bar...".  could this also support `progress=True`?

**Review by tlambert03 — COMMENTED** [review]
this is looking great!  In general I like that API.  left a couple specific comments

File: examples/progress_bar_threading.py
Comment (DragaDoncila) [reply examples/progress_bar_threading.py]: Yeah you're right, not becoming indeterminate is just a less useful option. Fixed that now.

**Comment by DragaDoncila** [comment]
@tlambert03 I think this is ready for re-review. I've addressed your concerns and also added indeterminate progress bars for `FunctionWorkers`. Happy to roll that back if we think it's too big a change.


The test failure is:

```python
  __________________________ test_calc_data_fast_uint8 __________________________
  
      def test_calc_data_fast_uint8():
  >       data = da.random.randint(
              0,
              100,
              size=(100_000, 1000, 1000),
              chunks=(1, 1000, 1000),
              dtype=np.uint8,
          )
  
  napari\layers\utils\_tests\test_layer_utils.py:62: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  d:\tmp\.tox\py38-windows-pyqt\lib\site-packages\dask\array\random.py:382: in randint
      return self._wrap(
  d:\tmp\.tox\py38-windows-pyqt\lib\site-packages\dask\array\random.py:148: in _wrap
      seeds = random_state_data(len(sizes), self._numpy_state)
  d:\tmp\.tox\py38-windows-pyqt\lib\site-packages\dask\utils.py:293: in random_state_data
      random_data = random_state.bytes(624 * n * 4)  # `n * 624` 32-bit integers
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  
  >   ???
  E   numpy.core._exceptions._ArrayMemoryError: Unable to allocate 238. MiB for an array with shape (62400000,) and data type uint32
  
  mtrand.pyx:803: MemoryError
```

We've had this test for about a year so I'm not sure what could be going wrong here, or why it thinks the data dype is `uint32`, but it may be a recent `dask` change?

File: napari/_qt/qthreading.py
Comment (tlambert03) [inline napari/_qt/qthreading.py:525]: type hint above doesn't match type hint here

File: napari/_qt/qthreading.py
Comment (tlambert03) [inline napari/_qt/qthreading.py]: this could maybe be a warning?

**Review by tlambert03 — APPROVED** [review]
slick :)

**Comment by sofroniewn** [comment]
Changes, have been made, i think this is good to merge!

**Review by jni — APPROVED** [review]

**Pull request merged**
